plugins/chat/assets/javascripts/discourse/models/chat-message.js

```diff
@@ -1,26 +1,193 @@
+  static create(channel, args = {}) {
+    return new ChatMessage(channel, args);
+  }
 
-    return this._super(args);
-  },
+  static createStagedMessage(channel, args = {}) {
+    args.staged_id = guid();
+    return new ChatMessage(channel, args);
+  }
 
-  _initReactions(args) {
-    args.reactions = EmberObject.create(args.reactions || {});
-  },
+  @tracked id;
+  @tracked error;
+  @tracked selected;
+  @tracked channel;
+  @tracked stagedId;
+  @tracked channelId;
+  @tracked createdAt;
+  @tracked deletedAt;
+  @tracked uploads;
+  @tracked excerpt;
+  @tracked message;
+  @tracked threadId;
+  @tracked reactions;
+  @tracked reviewableId;
+  @tracked user;
+  @tracked cooked;
+  @tracked inReplyTo;
+  @tracked expanded;
+  @tracked bookmark;
+  @tracked userFlagStatus;
+  @tracked hidden;
+  @tracked version = 0;
+  @tracked edited;
+  @tracked chatWebhookEvent = new TrackedObject();
+  @tracked mentionWarning;
+  @tracked availableFlags;
+  @tracked newest = false;
 
-  _initUserModel(args) {
-    if (!args.user || args.user instanceof User) {
-      return;
+  constructor(channel, args = {}) {
+    this.channel = channel;
+    this.id = args.id;
+    this.newest = args.newest;
+    this.edited = args.edited;
+    this.availableFlags = args.available_flags;
+    this.hidden = args.hidden;
+    this.threadId = args.thread_id;
+    this.channelId = args.chat_channel_id;
+    this.chatWebhookEvent = args.chat_webhook_event;
+    this.createdAt = args.created_at;
+    this.deletedAt = args.deleted_at;
+    this.excerpt = args.excerpt;
+    this.reviewableId = args.reviewable_id;
+    this.userFlagStatus = args.user_flag_status;
+    this.inReplyTo = args.in_reply_to
+      ? ChatMessage.create(channel, args.in_reply_to)
+      : null;
+    this.message = args.message;
+    this.cooked = args.cooked || ChatMessage.cookFunction(this.message);
+    this.reactions = this.#initChatMessageReactionModel(
+      args.id,
+      args.reactions
+    );
+    this.stagedId = args.staged_id;
+    this.uploads = new TrackedArray(args.uploads || []);
+    this.user = this.#initUserModel(args.user);
+    this.bookmark = args.bookmark ? Bookmark.create(args.bookmark) : null;
+  }
+
+  get read() {
+    return this.channel.currentUserMembership?.last_read_message_id >= this.id;
+  }
+
+  get firstMessageOfTheDayAt() {
+    if (!this.previousMessage) {
+      return this.#calendarDate(this.createdAt);
     }
 
-    args.user = User.create(args.user);
-  },
-});
+    if (
+      !this.#areDatesOnSameDay(
+        new Date(this.previousMessage.createdAt),
+        new Date(this.createdAt)
+      )
+    ) {
+      return this.#calendarDate(this.createdAt);
+    }
+  }
+
+  #calendarDate(date) {
+    return moment(date).calendar(moment(), {
+      sameDay: `[${I18n.t("chat.chat_message_separator.today")}]`,
+      lastDay: `[${I18n.t("chat.chat_message_separator.yesterday")}]`,
+      lastWeek: "LL",
+      sameElse: "LL",
+    });
+  }
+
+  @cached
```

Comment:
I did not realise that these `get`ters did not cache by default, I thought they were like computed properties and only changed when something it uses changes, e.g. `this.channel.messages`. Good to know.